#4345 API design: macOS and Windows OS vulnerabilities (#15492)
fleetdm/fleet: docs/REST API/rest-api.md

- [Get mobile device management (MDM) summary](#get-mobile-device-management-mdm-summary)
- [Get host's macadmin mobile device management (MDM) and Munki information](#get-hosts-macadmin-mobile-device-management-mdm-and-munki-information)
- [Get aggregated host's mobile device management (MDM) and Munki information](#get-aggregated-hosts-macadmin-mobile-device-management-mdm-and-munki-information)
- [List host OS versions](#list-host-os-versions)

> - [Get host OS version](#get-host-os-version)

Reviewer: @rachaelshaw seems like we're replacing `get`, should this line be removed?

Author: Before I delete a bunch of stuff, just want to double-check: we only ended up using `GET /api/v1/fleet/os_versions` and didn't need to add an endpoint for `GET /api/v1/fleet/os_versions/:id`?

Reviewer: Correct, we can use the same endpoint with a `name` and `version` filter to get the information on the osDetails page.  Also since we are [aggregating](https://github.com/fleetdm/fleet/issues/16234) operating systems, there's not a single id per operating system.

      "name_only": "Microsoft Windows 11 Pro 23H2",
      "version": "10.0.22621.1234",
      "platform": "windows",
      "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",
      "vulnerabilities": [
        {
          "cve": "CVE-2022-30190",
          "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",// Available in Fleet Premium
          "cvss_score": 7.8,// Available in Fleet Premium
          "epss_probability": 0.9729,// Available in Fleet Premium
          "cisa_known_exploit": false,// Available in Fleet Premium
          "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
          "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
          "resolved_in_version": ""// Available in Fleet Premium
        }
      ]
    }
  ],
  "meta": {
    "has_next_results": false,
    "has_previous_results": false
  }
}
```

`generated_cpe` and `vulnerabilities` are currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

```json
{

>   "id": 321,

Author: ```suggestion
```

| platform            | string   | query | Filters the hosts to the specified platform |
| os_name     | string | query | The name of the operating system to filter hosts by. `os_version` must also be specified with `os_name`                                                 |
| os_version    | string | query | The version of the operating system to filter hosts by. `os_name` must also be specified with `os_version`                                                 |
| page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |
| per_page                | integer | query | Results per page.                                                                                                                                                          |

> | order_key               | string  | query | What to order results by. Allowed fields are `hosts_count`. Default is `hosts_count` (descending).      |

Author: ```suggestion
| order_key               | string  | query | What to order results by. Allowed fields are: `hosts_count`. Default is `hosts_count` (descending).      |
```

> | order_key               | string  | query | What to order results by. Allowed fields are: `hosts_count`. Default is `hosts_count` (descending).      |

Author: I realize this reads a little weird when there's only 1 sortable thing, but figured sticking to the format we use elsewhere makes it easy to document any sortable columns we add in the future.

| order_direction | string | query | **Requires `order_key`**. The direction of the order given the order key. Options include `asc` and `desc`. Default is `asc`. |

##### Default response

`Status: 200`

```json
{

>   "counts_updated_at": "2023-12-06T22:17:30Z",

Reviewer: @ghernandez345 does this look correct?  @rachaelshaw a total count is needed here for the frontend.

```suggestion
  "count": 1
  "counts_updated_at": "2023-12-06T22:17:30Z",
```

  "os_versions": [
    {

>       "id": 123,

Author: The previous documentation calls this `os_id`, but I didn't actually see `os_id` in the existing API response so not sure it actually existed. 
@mostlikelee let me know if this does need to be `os_id` instead of just `id`.

Reviewer: @rachaelshaw I mentioned the [aggregation](https://github.com/fleetdm/fleet/issues/16234) above, so we don't currently have a reliable `os_id` to display here.   Ok to remove it?

      "hosts_count": 21,

>       "name": "Microsoft Windows Server 2022 Datacenter 10.0.20348",

Author: ```suggestion
      "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",
```

>       "name_only": "Microsoft Windows Server 2022 Datacenter",

Author: ```suggestion
      "name_only": "Microsoft Windows 11 Pro 23H2",
```

>       "version": "10.0.20348.1234",

Author: ```suggestion
      "version": "10.0.22621.1234",
```

      "platform": "windows",

>       "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Reviewer: @rachaelshaw Windows won't be using a cpe to lookup vulnerabilities (it's using the MSRC feed instead of the NVD feed).  Do we still want to display a cpe here in that case, omit the field, or return an empty string?

Author: I think we usually omit the field if it's not applicable; should there be a different field in its place?

Reviewer: Windows looks up vulns by "name" and "version" which are already included here.  I'll omit the field for windows if that works.

Reviewer: since we're using windows as the example , it shouldn't include `generated_cpe`

          "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
          "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
          "resolved_in_version": ""// Available in Fleet Premium
        }
      ]
    }
  ],
  "meta": {
    "has_next_results": false,
    "has_previous_results": false
  }
}
```

`generated_cpe` and `vulnerabilities` are currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

```json
{
  "id": 321,
  "hosts_count": 1,
  "name": "CentOS Linux 7.9.2009",
  "name_only": "CentOS",
  "version": "7.9.2009",
  "platform": "rhel",
  "generated_cpe": "",
  "vulnerabilities": []
}
```

### Get host OS version

Retrieves information about the specified OS version.

`GET /api/v1/fleet/os_versions/:id`

#### Parameters

| Name | Type | In | Description |
| ---- | ---- | -- | ----------- |
| id   | integer | path | **Required.** The OS version's ID. |

##### Default response

`Status: 200`

```json
{
  "os_version": {
    "id": 123,
    "hosts_count": 21,

>     "name": "Microsoft Windows Server 2022 Datacenter 10.0.20348",

Author: ```suggestion
    "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",
```

>     "name_only": "Microsoft Windows Server 2022 Datacenter",

Author: ```suggestion
    "name_only": "Microsoft Windows 11 Pro 23H2",
```

>     "version": "10.0.20348.1234",

Author: ```suggestion
    "version": "10.0.22621.1234",
```

    "platform": "windows",

>     "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Reviewer: for macos we use 2 cpes.  if it's ok we need to convert this to an array of `generated_cpes: []`

    "vulnerabilities": [
      {
        "cve": "CVE-2022-30190",

>         "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",// Available in Fleet Premium

Reviewer: @rachaelshaw I believe `details_link` is free for host software vulns

```json
{
  "counts_updated_at": "2023-12-06T22:17:30Z",
  "os_versions": [
    {
      "hosts_count": 21,
      "name": "Microsoft Windows Server 2022 Datacenter 10.0.20348",
      "name_only": "Microsoft Windows Server 2022 Datacenter",

>       "version": "10.0.20348",

Reviewer: nit:  windows OS vulnerabilities seem to match on the full version which adds an additional version number:
```suggestion
      "version": "10.0.20348.1234",
```

      "platform": "windows",

>       "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:-:*:*:*:*:*:*:*",

Reviewer: to note, macos will use a generated_cpe for vulnerability matching against the NVD data, but windows currently pulls vulns from MSRC which does not use the generated_cpe.

Reviewer: the cpe will include the version number: 
```suggestion
      "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",
```

| os_version    | string | query | The version of the operating system to filter hosts by. `os_name` must also be specified with `os_version`                                                 |
| page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |
| per_page                | integer | query | Results per page.                                                                                                                                                          |
| order_key               | string  | query | What to order results by. Allowed fields are: `hosts_count`. Default is `hosts_count` (descending).      |
| order_direction | string | query | **Requires `order_key`**. The direction of the order given the order key. Options include `asc` and `desc`. Default is `asc`. |

##### Default response

`Status: 200`

```json
{
  "count": 1
  "counts_updated_at": "2023-12-06T22:17:30Z",
  "os_versions": [
    {
      "hosts_count": 21,

>       "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",

Reviewer: @rachaelshaw @RachelElysia IIRC we're adding in a `os_version_id` here?

      "name_only": "Microsoft Windows 11 Pro 23H2",
      "version": "10.0.22621.1234",
      "platform": "windows",
      "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",
      "vulnerabilities": [
        {
          "cve": "CVE-2022-30190",

>           "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",// Available in Fleet Premium

Reviewer: @rachaelshaw this is a free field in `/hosts`

>           "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",// Available in Fleet Premium

Author: ```suggestion
          "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",
```

| platform            | string   | query | Filters the hosts to the specified platform |
| os_name     | string | query | The name of the operating system to filter hosts by. `os_version` must also be specified with `os_name`                                                 |
| os_version    | string | query | The version of the operating system to filter hosts by. `os_name` must also be specified with `os_version`                                                 |
| page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |
| per_page                | integer | query | Results per page.                                                                                                                                                          |
| order_key               | string  | query | What to order results by. Allowed fields are: `hosts_count`. Default is `hosts_count` (descending).      |

> | order_direction | string | query | **Requires `order_key`**. The direction of the order given the order key. Options include `asc` and `desc`. Default is `asc`. |

Reviewer: @rachaelshaw wasn't the old default `desc`?

          "epss_probability": 0.9729,// Available in Fleet Premium
          "cisa_known_exploit": false,// Available in Fleet Premium
          "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
          "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
          "resolved_in_version": ""// Available in Fleet Premium
        }
      ]
    }
  ],
  "meta": {
    "has_next_results": false,
    "has_previous_results": false
  }
}
```

`generated_cpe` and `vulnerabilities` are currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

```json
{
  "hosts_count": 1,
  "name": "CentOS Linux 7.9.2009",
  "name_only": "CentOS",
  "version": "7.9.2009",
  "platform": "rhel",
  "generated_cpe": "",
  "vulnerabilities": []
}
```

### Get host OS version

Retrieves information about the specified OS version.

`GET /api/v1/fleet/os_versions/:id`

#### Parameters

| Name | Type | In | Description |
| ---- | ---- | -- | ----------- |
| id   | integer | path | **Required.** The OS version's ID. |

##### Default response

`Status: 200`

```json

> {

Reviewer: we should still include `counts_updated_at` because the `hosts_count` is updated on a schedule.
```suggestion
{
  "counts_updated_at": "2023-12-06T22:17:30Z",
```

  "os_version": {
    "id": 123,
    "hosts_count": 21,
    "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",
    "name_only": "Microsoft Windows 11 Pro 23H2",
    "version": "10.0.22621.1234",
    "platform": "windows",

>     "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Author: ```suggestion
    "generated_cpes": [
      "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*"
    ],
```

Reviewer: yes indeed, but since the example here is windows, we omit the `generated_cpe`.  That field is only applicable to macOS.

    "vulnerabilities": [
      {
        "cve": "CVE-2022-30190",

>         "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",// Available in Fleet Premium

Author: ```suggestion
        "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",
```

| page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |
| per_page                | integer | query | Results per page.                                                                                                                                                          |
| order_key               | string  | query | What to order results by. Allowed fields are: `hosts_count`. Default is `hosts_count` (descending).      |
| order_direction | string | query | **Requires `order_key`**. The direction of the order given the order key. Options include `asc` and `desc`. Default is `asc`. |

##### Default response

`Status: 200`

```json
{
  "count": 1
  "counts_updated_at": "2023-12-06T22:17:30Z",
  "os_versions": [
    {
      "os_version_id": 123,
      "hosts_count": 21,
      "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",
      "name_only": "Microsoft Windows 11 Pro 23H2",
      "version": "10.0.22621.1234",
      "platform": "windows",

>       "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Author: ```suggestion
      "generated_cpes": [],
```

      "vulnerabilities": [
        {
          "cve": "CVE-2022-30190",
          "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",
          "cvss_score": 7.8,// Available in Fleet Premium
          "epss_probability": 0.9729,// Available in Fleet Premium
          "cisa_known_exploit": false,// Available in Fleet Premium
          "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
          "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
          "resolved_in_version": ""// Available in Fleet Premium
        }
      ]
    }
  ],
  "meta": {
    "has_next_results": false,
    "has_previous_results": false
  }
}
```

> `generated_cpe` and `vulnerabilities` are currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

Author: ```suggestion
Vulnerability data is currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:
```

```json
{
  "hosts_count": 1,
  "name": "CentOS Linux 7.9.2009",
  "name_only": "CentOS",
  "version": "7.9.2009",
  "platform": "rhel",

>   "generated_cpe": "",

Author: ```suggestion
  "generated_cpes": [],
```

  "vulnerabilities": []
}
```

### Get host OS version

Retrieves information about the specified OS version.

`GET /api/v1/fleet/os_versions/:id`

#### Parameters

| Name | Type | In | Description |
| ---- | ---- | -- | ----------- |
| id   | integer | path | **Required.** The OS version's ID. |

##### Default response

`Status: 200`

```json
{
  "os_version": {
    "id": 123,
    "hosts_count": 21,
    "name": "Microsoft Windows 11 Pro 23H2 10.0.22621.1234",
    "name_only": "Microsoft Windows 11 Pro 23H2",
    "version": "10.0.22621.1234",
    "platform": "windows",

>     "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Author: ```suggestion
    "generated_cpe": [],
```

>     "generated_cpe": "cpe:2.3:o:microsoft:windows_server_2022:10.0.20348.1234:*:*:*:*:*:*:*",

Author: ```suggestion
    "generated_cpes": [],
```

    "vulnerabilities": [
      {
        "cve": "CVE-2022-30190",
        "details_link": "https://nvd.nist.gov/vuln/detail/CVE-2022-30190",
        "cvss_score": 7.8,// Available in Fleet Premium
        "epss_probability": 0.9729,// Available in Fleet Premium
        "cisa_known_exploit": false,// Available in Fleet Premium
        "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
        "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
        "resolved_in_version": ""// Available in Fleet Premium
      }
    ]
  }
}
```

> `generated_cpe` and `vulnerabilities` are currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

Author: ```suggestion
OS vulnerability data is currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:
```

```json
{
  "id": 321,
  "hosts_count": 1,
  "name": "CentOS Linux 7.9.2009",
  "name_only": "CentOS",
  "version": "7.9.2009",
  "platform": "rhel",

>   "generated_cpe": "",

Author: ```suggestion
  "generated_cpes": [],
```

| platform            | string   | query | Filters the hosts to the specified platform |
| os_name     | string | query | The name of the operating system to filter hosts by. `os_version` must also be specified with `os_name`                                                 |

> | os_version    | string | query | The version of the operating system to filter hosts by. `os_name` must also be specified with `os_version`                                                 |

Reviewer: we also support `team_id`

Reviewer 2: Are we leaving `os_name` and `os_version` filter on here or is it being replaced with `os_version_id`?

Reviewer: I'll defer to @rachaelshaw, but if we remove them it'll be a breaking change

> | page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |

Author: ```suggestion
| team_id                | integer | query | _Available in Fleet Premium_. Filters to only include OS versions for the specified team.                                                                                                                                 |
| page                    | integer | query | Page number of the results to fetch.                                                                                                                                       |
```

          "cvss_score": 7.8,// Available in Fleet Premium
          "epss_probability": 0.9729,// Available in Fleet Premium
          "cisa_known_exploit": false,// Available in Fleet Premium
          "cve_published": "2022-06-01T00:15:00Z",// Available in Fleet Premium
          "cve_description": "Microsoft Windows Support Diagnostic Tool (MSDT) Remote Code Execution Vulnerability.",// Available in Fleet Premium
          "resolved_in_version": ""// Available in Fleet Premium
        }
      ]
    }
  ],
  "meta": {
    "has_next_results": false,
    "has_previous_results": false
  }
}
```

OS vulnerability data is currently available for Windows and macOS. For other platforms, `vulnerabilities` will be an empty array:

```json
{
  "hosts_count": 1,
  "name": "CentOS Linux 7.9.2009",
  "name_only": "CentOS",
  "version": "7.9.2009",
  "platform": "rhel",
  "generated_cpes": [],
  "vulnerabilities": []
}
```

### Get host OS version

Retrieves information about the specified OS version.

`GET /api/v1/fleet/os_versions/:id`

#### Parameters

| Name | Type | In | Description |
| ---- | ---- | -- | ----------- |

> | id   | integer | path | **Required.** The OS version's ID. |

Reviewer: need to support `team_id` here so we return the correct counts for that team

Reviewer: no hosts, so no need here

Reviewer: @rachaelshaw looks like the software detail page doesn't have a team context, so to keep things consistent we shouldn't need a `team_id` here, which means the OS Detail page will always have a global host count.

| software_version_id     | integer | query | The ID of the software version to filter hosts by.                                                                                                            |
| software_title_id       | integer | query | The ID of the software title to filter hosts by.                                                                                                              |

> | os_id                   | integer | query | The ID of the operating system to filter hosts by.                                                                                                                                                                                                                                                                                          |

Reviewer 2: Should `os_id` be removed?

> | os_id                   | integer | query | The ID of the operating system to filter hosts by.                                                                                                                                                                                                                                                                                          |

Author: ```suggestion
```
